### PR TITLE
PPL Integration - Add implementation for KMeans algorithm

### PR DESCRIPTION
**Pull request by jackiehanyang**
Signed-off-by: Jackie Han <hnyng@amazon.com>

### Description
[Describe what this change achieves]
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by jackiehanyang** [comment]
`build-windows32` and `build-windows64` check failed due to https://github.com/actions/virtual-environments/issues/4856

File: ppl/src/main/antlr/OpenSearchPPLParser.g4
Comment (penghuo) [inline ppl/src/main/antlr/OpenSearchPPLParser.g4:23]: Have considered more generic command name for ML? e.g. apply model-name?

File: core/build.gradle
Comment (penghuo) [inline core/build.gradle:50]: Core module should not depend on specific storage engine.

File: core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java:39]: Could you all MLCommonsOperator to opensearch module and add visitMLCommons() method in OpenSearchDefaultImplementor?

File: core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java:148]: Could you use ExprTulpleValue::value()?

File: core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java:59]: Does it make sense to parse Argument in Analyzer?

**Comment by penghuo** [comment]
Thanks for making the change!. Not finished yet, two high level comments
1. In general, Core module should know nothing about Storage Engine. Which means each storage engine should decide how to convert LogicalPlan to PhysicalPlan. So, MLCommonsOperator should be implemented in OpenSearch module instead of Core module.
2. Could you add doc for ML command also? e.g. https://github.com/penghuo/os-sql/blob/issue-396-pr/docs/user/ppl/index.rst

File: core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java
Comment (ylwu-amzn) [inline core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java:146]: MLCommons doesn't accept non-numerical data to run Kmeans. How about check if the value is numerical or not first? Maybe just choose numerical columns to pass to KMeans. Not so urgent, you can do this in next PR. Add some todo now is ok.

File: core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java
Comment (ylwu-amzn) [inline core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java:126]: MLCommons supports more data type now, how about adding `short`, `long` and `float` ?

File: ppl/src/main/java/org/opensearch/sql/ppl/utils/ArgumentFactory.java
Comment (ylwu-amzn) [inline ppl/src/main/java/org/opensearch/sql/ppl/utils/ArgumentFactory.java:147]: We can support other 2 parameters for Kmeans : `iterations` and `distanceType` , refer to this code https://github.com/opensearch-project/ml-commons/blob/main/common/src/main/java/org/opensearch/ml/common/parameter/KMeansParams.java#L39-L43

File: core/build.gradle
Comment (jackiehanyang) [reply core/build.gradle:50]: Yes it shouldn't be here, I forget to clear it here.

File: core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java
Comment (jackiehanyang) [reply core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java:39]: move all `MLCommonsOperator` to opensearch module? Sure will try this approach in next revision

File: core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java
Comment (jackiehanyang) [reply core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java:59]: I prefer to put Argument parsing logic in MLCommonsOperator, because I don't see any argument parsing logic in Analyzer for other logical plans. Also, it looks like the main purpose of Analyzer class is to construct the logical plan, so I prefer to leave the argument parsing work for the actual operator, which is MLCommonsOperator.

File: core/build.gradle
Comment (jackiehanyang) [reply core/build.gradle:50]: Attempted to move ml-commons dependency and opensearch dependency to opensearch module. However, it causing jar hell exception while installing opensearch1.3.

File: core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java
Comment (jackiehanyang) [reply core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java]: If we parse Argument in Analyzer, it will create dependencies in Core module as Analyzer class sits in Core module.

File: core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java
Comment (jackiehanyang) [reply core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java]: Added these three types in next revision

File: core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java
Comment (jackiehanyang) [reply core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java]: Added a check to only pass numerical columns to Kmeans in the next revision

File: core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java
Comment (jackiehanyang) [reply core/src/main/java/org/opensearch/sql/planner/physical/MLCommonsOperator.java]: I'm not quite sure what you mean. I don't think we are about to reference ExprTulpleValue here? Could you elaborate more?

File: ppl/src/main/antlr/OpenSearchPPLParser.g4
Comment (jackiehanyang) [reply ppl/src/main/antlr/OpenSearchPPLParser.g4:23]: We discussed how to construct the command so that it can be as generic as possible, and it seems like just pass in the algorithm name is the best solution. Model-name or model-id is hidden in the whole train and predict process. We don't want customers to make an effort to take notes on model attribute and pass it in the command.

File: core/build.gradle
Comment (ylwu-amzn) [reply core/build.gradle:50]: Will this PR https://github.com/opensearch-project/ml-commons/pull/137 fix the jar hell issue?

Comment (jackiehanyang) [reply core/build.gradle:50]: yes, after it's merged, I no longer see the jar hell issue in build.

**Comment by jackiehanyang** [comment]
Will send out a separate PR for documentation update along with documentation for AD integration.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/MLCommonsOperator.java
Comment (jackiehanyang) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/MLCommonsOperator.java:49]: Why `MLCommonsOperator` needs to be wrapped by `OpenSearchExecutionProtector`? extends two classes will make nesting structure.

File: ppl/src/main/java/org/opensearch/sql/ppl/utils/ArgumentFactory.java
Comment (ylwu-amzn) [inline ppl/src/main/java/org/opensearch/sql/ppl/utils/ArgumentFactory.java:165]: We already have these default values in MLCommons. How about we just set these values as null in PPL, that means we just let MLCommons to decide what's the default value. It may conflict in future if we have both PPL and MLCommons set default values.

File: ppl/src/main/antlr/OpenSearchPPLLexer.g4
Comment (ylwu-amzn) [inline ppl/src/main/antlr/OpenSearchPPLLexer.g4:271]: Seems `DOUBLE_LITERAL` is same with `DECIMAL_LITERAL`. Do we really need this?

File: opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/ADOperator.java
Comment (ylwu-amzn) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/ADOperator.java:141]: Minor: duplicate code with `MLCommonsOperator`, can refactor this part.

File: opensearch/build.gradle
Comment (penghuo) [inline opensearch/build.gradle:41]: Does it required?

**Review by penghuo — APPROVED** [review]
Thanks for the change!

**Pull request merged**
